opengovsg/postmangovsg
# Migrate worker to get environment variables from Secrets Manager

## Problem

Worker uses environment variables set in the ECS task definition, many of of them are similar to backend en vars. The worker can share secrets with backend and all the secrets can be consolidated in a secret in Secret Manager. Hence, we need to migrate worker to pull secrets from Secrets Manager too. There are some conflicting env vars which have similar names as backend en vars, but differ in value - these have to be renamed.

Closes #813 

## Solution

**Features**:

- Pull secrets from Secret Manager 
- Load secrets in .env file before starting up worker

## Tests

- Deploy to staging
- Create a new task definition by removing e.g. `DB_URI` env var from the previous ECS task definition
- Create a campaign and ensure campaign is sent out successfully

## Deploy Notes

**New environment variables**:

Renamed `worker` en vars to prepend `Worker`
- `SES_PASS` -> `WORKER_SES_HOST`
- `SES_PASS` -> `WORKER_SES_PASS`
- `SES_USER` -> `WORKER_SES_USER`
- `SES_PORT` -> `WORKER_SES_PORT`

**New scripts**:

- `before-deploy.sh` : Script that executes before travis deployment to inject secret id according to staging/prod branch
- `docker-entrypoint.sh`: Retrieve secrets from aws secrets manager and store them `.env` file

**Environment variables to be moved to Secret Manager**
- `AWS_REGION`
- `BACKEND_URL`
- `DB_URI`
- `DB_USE_IAM`
- `DEFAULT_COUNTRY`
- `NODE_ENV`
- `SECRET_MANAGER_SALT`
- `WORKER_SES_FROM`
- `WORKER_SES_HOST`
- `WORKER_SES_PASS`
- `WORKER_SES_PORT`
- `WORKER_SES_USER`
- `TWILIO_CALLBACK_SECRET`
- `UNSUBSCRIBE_HMAC_ALGO_V1`
- `UNSUBSCRIBE_HMAC_KEY_V1`

**Environment variables that will remain in ECS**
- `ECS_SERVICE_NAME`
- `MESSAGE_WORKER_LOGGER`
- `MESSAGE_WORKER_SENDER`
